File: backend/springboot/src/main/java/tn/maiko26/springboot/service/ProfileService.java
package tn.maiko26.springboot.service;

import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import tn.maiko26.springboot.exception.CustomException;
import tn.maiko26.springboot.model.User;
import tn.maiko26.springboot.model.UserSetting;
import tn.maiko26.springboot.repository.UserRepository;
import tn.maiko26.springboot.repository.UserSettingRepository;

@Service
public class ProfileService {

    private final UserService userService;
    private final UserRepository userRepository;
    private final UserSettingRepository userSettingRepository;
    private final PasswordEncoder passwordEncoder;

    public ProfileService(UserService userService, UserRepository userRepository, UserSettingRepository userSettingRepository) {
        this.userService = userService;
        this.userRepository = userRepository;
        this.userSettingRepository = userSettingRepository;
        this.passwordEncoder = new BCryptPasswordEncoder();
    }


    public User getCurrentUser() {
        String email = userService.getCurrentUserEmail();
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new CustomException("User not found", 400));
    }


    public User updateProfile(String name, String phoneNumber, String avatarPath) {
        User user = getCurrentUser();

        if (name != null) user.setName(name);
        if (phoneNumber != null) user.setPhoneNumber(phoneNumber);
        if (avatarPath != null) user.setAvatarUrl(avatarPath);

        return userRepository.save(user);
    }


    public void updatePassword(String currentPassword, String newPassword) {
        User user = getCurrentUser();

        if (!passwordEncoder.matches(currentPassword, user.getPassword())) {
            throw new CustomException("Current password is incorrect", 400);
        }

        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }


    public void updateNotificationSettings(UserSetting newSetting) {
        User user = getCurrentUser();
        UserSetting currentSetting = userSettingRepository.findByUser(user)
                .orElse(new UserSetting(user));


        if (newSetting.getEmailDigest() != null) {
            currentSetting.setEmailDigest(newSetting.getEmailDigest());
        }
        if (newSetting.getTaskReminders() != null) {
            currentSetting.setTaskReminders(newSetting.getTaskReminders());
        }

        userSettingRepository.save(currentSetting);
    }
}

File: backend/dotnet/Services/ProfileService.cs
using dotnet.Data;
using dotnet.exceptions;
using dotnet.Interfaces;
using dotnet.Models;
using Microsoft.AspNetCore.Identity;

namespace dotnet.Services;

public class ProfileService : IProfileService
{
    
     private readonly DataContext _context;
   private readonly IUserService _userService;
    private readonly HashPasswordService _hashPasswordUtil;

    public ProfileService(DataContext context, IUserService userService, HashPasswordService hashPasswordUtil)
   {
       _context = context;
       _userService = userService;
        _hashPasswordUtil = hashPasswordUtil;

    }

    public User GetCurrentUser()
   {
       var email = _userService.GetCurrentUserEmail();
       return _context.Users.FirstOrDefault(u => u.Email == email) ?? 
              throw new CustomException("User not found", 400);
   }

   public User UpdateProfile(string name, string phoneNumber, string avatarPath)
   {
       var user = GetCurrentUser();

       if (name != null) user.Name = name;
       if (phoneNumber != null) user.PhoneNumber = phoneNumber;
       if (avatarPath != null) user.AvatarUrl = avatarPath;

       _context.Update(user);
       _context.SaveChanges();
       return user;
   }

   public void UpdatePassword(string currentPassword, string newPassword)
   {
       var user = GetCurrentUser();

       var passwordVerification = _hashPasswordUtil.VerifyPassword(user.Password, currentPassword);
       if (!passwordVerification)
       {
           throw new CustomException("Current password is incorrect", 400);
       }

       user.Password = _hashPasswordUtil.HashPassword(newPassword);
       _context.Update(user);
       _context.SaveChanges();
   }

   public void UpdateNotificationSettings(UserSetting newSetting)
   {
       var user = GetCurrentUser();
       var currentSetting = _context.UserSettings.FirstOrDefault(s => s.User == user) ?? 
                           new UserSetting { User = user };

       if (newSetting.EmailDigest != null)
           currentSetting.EmailDigest = newSetting.EmailDigest;
       if (newSetting.TaskReminders != null)
           currentSetting.TaskReminders = newSetting.TaskReminders;

       _context.UserSettings.Update(currentSetting);
       _context.SaveChanges();
   }
}